File: references/betterlectio/ios/BetterLectio/MessageModels.swift
//
//  MessageModels.swift
//  BetterLectio
//
//  Created by Elliott Friedrich on 03/02/2026.
//

import Foundation

// MARK: - Message Models

/// Represents a message thread in the inbox
struct MessageThread: Codable, Identifiable, Equatable, Hashable, Sendable {
    let id: String                // thread ID from __doPostBack argument
    let title: String             // message subject
    let senderName: String        // name of latest sender
    let firstSenderName: String   // name of original sender
    let recipients: String        // recipient group(s)
    let date: String              // formatted date string
    let isRead: Bool              // read status
    let isFlagged: Bool           // flagged/starred status
    let hasAttachment: Bool       // has file attachment
    let senderType: SenderType    // student or teacher

    func with(isRead: Bool? = nil, isFlagged: Bool? = nil) -> MessageThread {
        MessageThread(
            id: id,
            title: title,
            senderName: senderName,
            firstSenderName: firstSenderName,
            recipients: recipients,
            date: date,
            isRead: isRead ?? self.isRead,
            isFlagged: isFlagged ?? self.isFlagged,
            hasAttachment: hasAttachment,
            senderType: senderType
        )
    }

    func markedRead() -> MessageThread { with(isRead: true) }
    func toggledRead() -> MessageThread { with(isRead: !isRead) }
    func toggledFlag() -> MessageThread { with(isFlagged: !isFlagged) }
}

enum SenderType: String, Codable, Sendable {
    case student
    case teacher
    case unknown
}

/// Represents a single message within a thread
struct Message: Codable, Identifiable, Equatable, Sendable {
    let id: String
    let senderName: String
    let date: String              // formatted date with time
    let title: String
    let content: String           // HTML content of the message
    let editedAt: Date?
    let attachments: [MessageAttachment]
    let senderEntityID: String?
    let locator: MessageLocator?
    let reactions: [MessageReactionGroup]
    let ownReaction: MessageReactionEmoji?
    let editPostbackTarget: String

    init(
        id: String,
        senderName: String,
        date: String,
        title: String,
        content: String,
        editedAt: Date? = nil,
        attachments: [MessageAttachment],
        senderEntityID: String? = nil,
        locator: MessageLocator? = nil,
        reactions: [MessageReactionGroup] = [],
        ownReaction: MessageReactionEmoji? = nil,
        editPostbackTarget: String = ""
    ) {
        self.id = id
        self.senderName = senderName
        self.date = date
        self.title = title
        self.content = content
        self.editedAt = editedAt
        self.attachments = attachments
        self.senderEntityID = senderEntityID
        self.locator = locator
        self.reactions = reactions
        self.ownReaction = ownReaction
        self.editPostbackTarget = editPostbackTarget
    }

    private enum CodingKeys: String, CodingKey {
        case id, senderName, date, title, content, editedAt, attachments, senderEntityID, locator, reactions, ownReaction, editPostbackTarget
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decode(String.self, forKey: .id)
        senderName = try container.decode(String.self, forKey: .senderName)
        date = try container.decode(String.self, forKey: .date)
        title = try container.decode(String.self, forKey: .title)
        content = try container.decode(String.self, forKey: .content)
        editedAt = try container.decodeIfPresent(Date.self, forKey: .editedAt)
        attachments = try container.decode([MessageAttachment].self, forKey: .attachments)
        senderEntityID = try container.decodeIfPresent(String.self, forKey: .senderEntityID)
        locator = try container.decodeIfPresent(MessageLocator.self, forKey: .locator)
        reactions = try container.decodeIfPresent([MessageReactionGroup].self, forKey: .reactions) ?? []
        ownReaction = try container.decodeIfPresent(MessageReactionEmoji.self, forKey: .ownReaction)
        editPostbackTarget = try container.decodeIfPresent(String.self, forKey: .editPostbackTarget) ?? ""
    }

    func with(
        locator: MessageLocator?,
        reactions: [MessageReactionGroup],
        ownReaction: MessageReactionEmoji?
    ) -> Message {
        Message(
            id: id,
            senderName: senderName,
            date: date,
            title: title,
            content: content,
            editedAt: editedAt,
            attachments: attachments,
            senderEntityID: senderEntityID,
            locator: locator,
            reactions: reactions,
            ownReaction: ownReaction,
            editPostbackTarget: editPostbackTarget
        )
    }
}

struct MessageEditDraft: Equatable, Sendable {
    let locator: MessageLocator
    let title: String
    let body: String
    let signatureSuffix: String
    let editHTML: String
    let formAction: String
    let titleField: String
    let bodyField: String
    let saveTarget: String
}

/// Represents a file attachment on a message
struct MessageAttachment: Codable, Identifiable, Equatable, Sendable {
    let id: String
    let name: String
    let url: String
}

/// Complete message thread with all messages
struct MessageThreadDetail: Codable, Equatable, Sendable {
    let threadId: String
    let title: String
    let recipients: String
    let recipientIDs: [String]
    let messages: [Message]
    let canReply: Bool

    init(
        threadId: String,
        title: String,
        recipients: String,
        recipientIDs: [String] = [],
        messages: [Message],
        canReply: Bool
    ) {
        self.threadId = threadId
        self.title = title
        self.recipients = recipients
        self.recipientIDs = recipientIDs
        self.messages = messages
        self.canReply = canReply
    }

    private enum CodingKeys: String, CodingKey {
        case threadId, title, recipients, recipientIDs, messages, canReply
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        threadId = try container.decode(String.self, forKey: .threadId)
        title = try container.decode(String.self, forKey: .title)
        recipients = try container.decode(String.self, forKey: .recipients)
        recipientIDs = try container.decodeIfPresent([String].self, forKey: .recipientIDs) ?? []
        messages = try container.decode([Message].self, forKey: .messages)
        canReply = try container.decode(Bool.self, forKey: .canReply)
    }

    func withMessages(_ messages: [Message]) -> MessageThreadDetail {
        MessageThreadDetail(
            threadId: threadId,
            title: title,
            recipients: recipients,
            recipientIDs: recipientIDs,
            messages: messages,
            canReply: canReply
        )
    }
}

/// Form data for sending a reply
struct MessageReplyForm: Codable, Sendable {
    let title: String
    let content: String
    let notifyOption: NotifyOption
    let attachmentDocId: String?
}

enum NotifyOption: String, Codable, Sendable {
    case senderOnly           // notify only the original sender
    case none                 // notify nobody
    case all                  // notify all recipients
}

// MARK: - Message Folder

/// Represents a Lectio message folder. IDs come straight from Lectio's
/// `ListGridSelectionTree$folders` field — either the standard virtual folders
/// (Nyeste `-70`, Ulæst `-40`, Egne beskeder `-10`, Sendte `-80`, Slettede `-60`) or a
/// positive user-created folder ID.
struct MessageFolder: Codable, Hashable, Identifiable, Sendable {
    let id: String
    let displayName: String

    static let newest  = MessageFolder(id: "-70", displayName: "Nyeste")
    static let unread  = MessageFolder(id: "-40", displayName: "Ulæst")
    static let inbox   = MessageFolder(id: "-10", displayName: "Egne beskeder")
    static let sent    = MessageFolder(id: "-80", displayName: "Sendte beskeder")
    static let deleted = MessageFolder(id: "-60", displayName: "Alle slettede")

    /// Seed list shown before the first HTML parse completes and as an offline fallback.
    static let defaults: [MessageFolder] = [.newest, .unread, .inbox, .sent, .deleted]
}

// MARK: - Message Composer Model

struct MessageRecipient: Codable, Identifiable, Equatable, Hashable, Sendable {
    let id: String
    let name: String
    let type: RecipientType
    let info: String?           // class, abbreviation, etc.
    let lectioRecipientID: String
}

enum RecipientType: String, Codable, Sendable {
    case student
    case teacher
    case group
    case hold
}

File: references/betterlectio/android/app/src/main/java/dk/betterlectio/android/feature/messages/MessageModels.kt
package dk.betterlectio.android.feature.messages

import android.net.Uri
import java.time.LocalDateTime
import java.time.Instant

data class MessageFolder(
    val id: String,
    val displayName: String,
) {
    companion object {
        val NEWEST = MessageFolder("-70", "Nyeste")
        val UNREAD = MessageFolder("-40", "Ulæst")
        val INBOX = MessageFolder("-10", "Egne beskeder")
        val SENT = MessageFolder("-80", "Sendte beskeder")
        val DELETED = MessageFolder("-60", "Alle slettede")
        val defaults = listOf(NEWEST, UNREAD, INBOX, SENT, DELETED)
    }
}

data class MessageThread(
    val id: String,
    val topic: String,
    /** Latest sender display name (title attribute when present). */
    val sender: String,
    val dateChanged: LocalDateTime?,
    val folderId: String,
    val normalizedId: String = id,
    val unread: Boolean = false,
    val flagged: Boolean = false,
    /** Lectio context-card id when present on the list row (`S…` / `T…`). */
    val senderEntityId: String? = null,
    /** `STUDENT` / `TEACHER` when inferred from list row fonticon classes. */
    val senderKind: String? = null,
)

data class MessageAttachment(
    val name: String,
    val url: String,
)

enum class MessageReactionEmoji(val glyph: String) {
    THUMBS_UP("👍"),
    HEART("❤️"),
    LAUGH("😂"),
    SURPRISED("😮"),
    SAD("😢"),
    THUMBS_DOWN("👎");

    companion object {
        fun fromGlyph(value: String?): MessageReactionEmoji? = entries.firstOrNull { it.glyph == value }
    }
}

data class MessageLocator(
    val senderKey: String,
    val sentAt: String,
    val occurrence: Int,
)

data class MessageReactionParticipant(
    val key: String,
    val name: String,
    val isOwn: Boolean,
)

data class MessageReactionGroup(
    val emoji: MessageReactionEmoji,
    val reactors: List<MessageReactionParticipant>,
)

data class ThreadEntry(
    val id: String,
    val topic: String?,
    val contentHtml: String?,
    val senderName: String?,
    val sentAt: LocalDateTime?,
    val editedAt: Instant? = null,
    val attachments: List<MessageAttachment> = emptyList(),
    val senderEntityId: String? = null,
    val senderKind: String? = null,
    val locator: MessageLocator? = null,
    val reactions: List<MessageReactionGroup> = emptyList(),
    val ownReaction: MessageReactionEmoji? = null,
    /** Present only when Lectio grants edit permission for this row. */
    val editPostbackTarget: String = "",
) {
    val attachmentNames: List<String> get() = attachments.map { it.name }
}

data class MessageEditDraft(
    val thread: MessageThread,
    val locator: MessageLocator,
    val title: String,
    val body: String,
    val signatureSuffix: String,
    internal val editHtml: String,
    internal val formAction: String,
    internal val titleField: String,
    internal val bodyField: String,
    internal val saveTarget: String,
)

data class MessageThreadDetail(
    val thread: MessageThread,
    val entries: List<ThreadEntry>,
    val receivers: List<String> = emptyList(),
    /** Lectio context-card ids for thread recipients (`S…` / `T…`) — signature skip. */
    val receiverEntityIds: List<String> = emptyList(),
    /** Fresh row-scoped edit targets for the signed-in user's reaction carriers. */
    val ownReactionCarrierTargets: Map<MessageLocator, String> = emptyMap(),
)

/**
 * Local attachment selected in compose/reply UI (uploaded at send time).
 */
data class ComposeAttachment(
    val uri: Uri,
    val displayName: String,
    val mimeType: String,
    val sizeBytes: Long? = null,
)

data class ComposeMessageDraft(
    val subject: String,
    /** BBCode body (Lectio `EditModeContentBBTB`). */
    val body: String,
    val recipientIds: List<String>,
    val recipientNames: List<String>,
    /** When true, inject Lectio `RepliesNotAllowedChkBox=on` on compose postbacks. */
    val repliesNotAllowed: Boolean = false,
    val attachments: List<ComposeAttachment> = emptyList(),
)

data class MessageRecipient(
    val id: String,
    val name: String,
    val kind: String = "person",
)

/**
 * Resolved attachment field names from a compose/reply form HTML page.
 */
data class MessageAttachTargets(
    val docIdFieldName: String,
    val postbackTarget: String,
)
